src/display_list/list.rs

```diff
@@ -0,0 +1,104 @@
+use super::annotation::Annotation;
+use super::line::{DisplayLine, DisplayRawLine, DisplaySourceLine};
+use crate::slice::Slice;
+use std::fmt;
+
+#[derive(Debug, Clone)]
+pub struct DisplayList<'d> {
+    pub body: Vec<DisplayLine<'d>>,
+}
+
+fn get_header_pos(slice: &Slice) -> (Option<usize>, Option<usize>) {
+    let line = slice.line_start;
+    (line, None)
+}
+
+impl<'d> From<&Slice<'d>> for DisplayList<'d> {
+    fn from(slice: &Slice<'d>) -> Self {
+        let mut body = vec![];
```

Comment:
why allocate all this and not just render it lazily in the `Display` impl?

Reply:
The reason to separate building displaylist from formatting (`Display`) is that we need to compute the layout of the output before we draw it.

If I understand your question, you're saying that we could just take the `Slice/Snippet` and write the output, but due to the nature of the library, we need to perform a lot of computations to layout the structure where the decisions about the first line depend on the information in the last line etc.

Therefore, we need a "two pass" approach I believe. We can combine those two passes in one method, but then we limit ourselves in several ways.

My take on the `Slice -> DisplayList -> Output` is that it allows me to design a layout builder (`Slice -> DisplayList`) and separately a "rendering" code (`DisplayList -> Output`).
In the simplest use case, we can differentiate between "themes" that will take the same `DisplayList` and draw it differently (color, no color, rich unicode characters to draw tables/frames etc.) and we can also have different layouts ('rustc-like`, `elm-like` etc.).

I understand that it looks like it introduces cost in form of allocation, and I'd like to look for ways to minimize it, but I believe the clean codebase, separation of concerns, and ability to customize those two aspects justify the logic.
Does it sound reasonable to you?

Reply:
Yes totally. We should document this though ;) I was mostly curious as to why this was happening.